File: ThemesViewController.m
//
//  ThemesViewController.m
//  Homework4
//
//  Created by Aleksandr Avdukich on 06/03/2019.
//  Copyright © 2019 Sanel Avdukich. All rights reserved.
//

#import "ThemesViewController.h"
#import "Themes.h"
#import "ThemesViewControllerDelegate.h"

@implementation ThemesViewController

-(id<ThemesViewControllerDelegate>)delegate {
  return _delegate;
}

-(Themes *)model {
  return _model;
}

- (IBAction)changeButton:(id)sender {
  
  UINavigationBar *bar = [self.navigationController navigationBar];
  
  switch ([sender tag])
  {
    case 1:
      [[self delegate] themesViewController:self didSelectTheme:[[self model] theme1] didSelectTint:[[self model] darkTint]]; // (lldb) po _delegate = _TtC10Homework_418RootViewController *    0x7f8c4fc13e10 0x00007f8c4fc13e10 или nil (0x0).
      [bar setBarTintColor: [[self model] theme1]];
      [bar setTintColor:[[self model] darkTint]];
      [bar setTitleTextAttributes:@{NSForegroundColorAttributeName:[[self model] darkTint]}];
      break;
    case 2:
      [[self delegate] themesViewController: self didSelectTheme:[[self model] theme2] didSelectTint:[[self model] whiteTint]];
      [bar setBarTintColor: [[self model] theme2]];
      [bar setTintColor:[[self model] whiteTint]];
      [bar setTitleTextAttributes:@{NSForegroundColorAttributeName:[[self model] whiteTint]}];
      break;
    case 3:
      [[self delegate] themesViewController: self didSelectTheme:[[self model] theme3] didSelectTint:[[self model] darkTint]];
      [bar setBarTintColor: [[self model] theme3]];
      [bar setTintColor:[[self model] darkTint]];
      [bar setTitleTextAttributes:@{NSForegroundColorAttributeName:[[self model] darkTint]}];
      break;
    default:
      break;
  }
  
}


- (IBAction)saveButton:(id)sender
{
  [self dismissViewControllerAnimated:YES completion:nil];
}

- (void)viewDidLoad {
  [super viewDidLoad];
  
  [[UINavigationBar appearance] setTintColor:UIColor.blackColor];
  
  self.model = [[Themes alloc] init];
}

-(void)setDelegate:(id<ThemesViewControllerDelegate>)delegate {
  if (delegate != _delegate) {
    id<ThemesViewControllerDelegate> oldDelegate = _delegate;
    _delegate = [delegate retain];
    [oldDelegate release];
  }
}

-(void)setModel:(Themes *)model {
  @synchronized(self) {
    _model = model;
  }
}

-(void) dealloc
{
  [[self model] release];
  [[self delegate] release];
  [super dealloc];
}



@end

File: Theme Swift/ThemesViewController.swift
//
//  ThemesViewController.swift
//  Homework4
//
//  Created by Aleksandr Avdukich on 06/03/2019.
//  Copyright © 2019 Sanel Avdukich. All rights reserved.
//

import UIKit

class ThemesViewController: UIViewController {
  
  private var model: ThemesSwift?
  var delegate: ThemesViewControllerDeleateSwift?
  var closure: (() -> ())?
  
  
  @IBAction func changeButton(_ sender: UIButton) {
    
    switch sender.tag {
    case 1:
      let theme = self.model?.theme1 ?? UIColor.black
      let tint = self.model?.darkTint ?? UIColor.black
      delegate?.themesViewController(didSelectTheme: theme, didSelectTint: tint)
      closure!()
      navigationController?.navigationBar.barTintColor = theme
      navigationController?.navigationBar.tintColor = tint
      navigationController?.navigationBar.titleTextAttributes = [NSAttributedString.Key.foregroundColor: tint]
      break
    case 2:
      let theme = self.model?.theme2 ?? UIColor.black
      let tint = self.model?.whiteTint ?? UIColor.white
      delegate?.themesViewController(didSelectTheme: theme, didSelectTint: tint)
      closure!()
      navigationController?.navigationBar.barTintColor = theme
      navigationController?.navigationBar.tintColor = tint
      navigationController?.navigationBar.titleTextAttributes = [NSAttributedString.Key.foregroundColor: tint]
      break
    case 3:
      let theme = self.model?.theme3 ?? UIColor.black
      let tint = self.model?.darkTint ?? UIColor.black
      delegate?.themesViewController(didSelectTheme: theme, didSelectTint: tint)
      closure!()
      navigationController?.navigationBar.barTintColor = theme
      navigationController?.navigationBar.tintColor = tint
      navigationController?.navigationBar.titleTextAttributes = [NSAttributedString.Key.foregroundColor: tint]
      break
    default:
      print("Wrong tag!")
    }
  }
  
  @IBAction func saveButton(_ sender: UIBarButtonItem) {
    self.dismiss(animated: true, completion: nil)
  }
  
  override func viewDidLoad() {
    super.viewDidLoad()
    
    navigationController?.navigationBar.tintColor = .black
    model = ThemesSwift()
    
  }

}
